### DISCUSSION TITLE
[2.0] DefaultExecutionContext.should_autocommit removed

### DISCUSSION THREAD

**Discussion by mathiasritter** [sqlalchemy/sqlalchemy#8854] [Usage Questions] [answered]
I'm using SQLAlchemy 2.0.0b3 with [trino](https://github.com/trinodb/trino-python-client). The trino dialect accesses `should_autocommit` on `DefaultExecutionContext`, which raises an error as the attribute doesn't exist:  https://github.com/trinodb/trino-python-client/blob/eef6388cd30f93ed1172cb1712b1b7e680276761/trino/sqlalchemy/dialect.py#L366

In the [1.4 docs](https://docs.sqlalchemy.org/en/14/core/internals.html#sqlalchemy.engine.default.DefaultExecutionContext), `should_autocommit` is still present, but in the [2.0 docs](https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.default.DefaultExecutionContext) it isn't.

What's the replacement for this attribute in 2.0?

UPDATE: Created an issue on the trino repository: https://github.com/trinodb/trino-python-client/issues/291

**Comment by CaselIT** [comment, marked as answer]
Hi,

The autocommit logic was removed for the 2.0 release, so there is no replacement.
Sadly the 1.4 documentation missed the deprecation notice.
@zzzeek should it also raise a warning at runtime?

**Reply by mathiasritter** [reply]
Hi, thanks for the quick answer. I don't know much about the implementation of this dialect (and dialects in general). Given that there is no replacement, what changes are necessary to migrate the dialect from 1.4 to 2.0? I might open an issue on the [trino](https://github.com/trinodb/trino-python-client) repository.

**Reply by zzzeek** [reply]
im -1 on a warning as this is a dialect-level attribute that isn't really accessed by the implementations themselves, we'd otherwise have to add every parameter on Dialect that was remove with warnings because we don't know which might be accessed for some unusual reason, IMO this is out of scope.

Dialects need to just make whatever minor changes are necessary to work on SQLAlchemy 2.0.    

in this case it looks like they are trying to detect DML/DDL, they should be using context.is_ddl / context.is_dml for that, the approach they have here is broken for any version of SQLAlchemy.  but also, ddl/dml will do a cursor.close() immediately as well , it seems quite unusual if one executed a DML/DDL on a cursor , then called cursor.close() without calling cursor.description, the statement wouldn't take place?    so not really sure why they need this at all and they should revisit what's going on here.

**Reply by hashhar** [reply]
Thanks for the tips @zzzeek . The trino-python-client in the past worked a bit weirdly - `cursor.execute` only submitted the query for execution but didn't guarantee query execution had started on the server. It was fixed later but apparently that SQLA dialect wasn't updated with it.

Seems we don't even need the DDL/DML check anymore and can just pass-through to `cursor.execute` now.
